Foundation/NSURLSession/HTTPMessage.swift

```diff
@@ -0,0 +1,367 @@
+// Foundation/NSURLSession/HTTPMessage.swift - HTTP Message parsing
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+// -----------------------------------------------------------------------------
+///
+/// Helpers for parsing HTTP responses.
+/// These are libcurl helpers for the NSURLSession API code.
+/// - SeeAlso: https://curl.haxx.se/libcurl/c/
+/// - SeeAlso: NSURLSession.swift
+///
+// -----------------------------------------------------------------------------
+
+import CoreFoundation
+import Dispatch
+
+
+
+extension URLSessionTask {
+    /// An HTTP header being parsed.
+    ///
+    /// It can either be complete (i.e. the final CR LF CR LF has been
+    /// received), or partial.
+    internal enum _ParsedResponseHeader {
+        case partial(_ResponseHeaderLines)
+        case complete(_ResponseHeaderLines)
+        init() {
+            self = .partial(_ResponseHeaderLines())
+        }
+    }
+    /// A type safe wrapper around multiple lines of headers.
+    ///
+    /// This can be converted into an `NSHTTPURLResponse`.
+    internal struct _ResponseHeaderLines {
+        let lines: [String]
+        init() {
+            self.lines = []
+        }
+        init(headerLines: [String]) {
+            self.lines = headerLines
+        }
+    }
+}
+
+extension URLSessionTask._ParsedResponseHeader {
+    /// Parse a header line passed by libcurl.
+    ///
+    /// These contain the <CRLF> ending and the final line contains nothing but
+    /// that ending.
+    /// - Returns: Returning nil indicates failure. Otherwise returns a new
+    ///     `ParsedResponseHeader` with the given line added.
+    @warn_unused_result
+    func byAppending(headerLine data: UnsafeBufferPointer<Int8>) -> URLSessionTask._ParsedResponseHeader? {
+        // The buffer must end in CRLF
+        guard
+            2 <= data.count &&
+                data[data.endIndex - 2] == CR &&
+                data[data.endIndex - 1] == LF
+            else { return nil }
+        let lineBuffer = UnsafeBufferPointer(start: UnsafePointer<UInt8>(data.baseAddress), count: data.count - 2)
+        guard let line = String(utf8Buffer: lineBuffer) else { return nil }
```

Comment:
Should be `init?(bytes:encoding:)` again here (and elsewhere we create strings) from data.